From 10fd8c23076fec2089dd602e3d0c2cb9ca60c64a Mon Sep 17 00:00:00 2001
From: lei zhilin <leizhilin@eswincomputing.com>
Date: Thu, 13 Feb 2025 11:10:23 +0800
Subject: [PATCH 312/534] fix(2d):kernel stack warning after reboot

Changelogs:

1.Problem cause: In gpu_parse_dt, devm_devfreq_add_device is opened and saved to the global variable gpd.
  However, in gpu_add_power_domains, memset is performed, which causes the saved node to be set to 0,
  resulting in a null pointer access exception.
2.Commit that introduced the problem: http://10.10.130.10:8080/c/win2030/linux-6.6/+/18265
3.Solutions:
  1) Move devm_devfreq_add_device into the gpu_add_power_domains function.
  2) Move devm_devfreq_remove_device into the gpu_remove_power_domains function.
  3) In the viv_dev_exit function, first call gckPLATFORM_Terminate to release the devfreq device,
     and then unregister the entire driver to prevent null pointer references.

Signed-off-by: lei zhilin <leizhilin@eswincomputing.com>
---
 .../os/linux/kernel/gc_hal_kernel_driver.c    |   2 +
 .../eswin/gc_hal_kernel_platform_win2030.c    | 146 ++++++++++++++----
 2 files changed, 122 insertions(+), 26 deletions(-)

diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_driver.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_driver.c
index eabfbfb1b2de..6c7ad73ea6f0 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_driver.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_driver.c
@@ -1556,6 +1556,7 @@ static int viv_dev_suspend(struct platform_device *dev, pm_message_t state)
                     return -1;
                 }
 
+                hae_print("hard: %p, store_state: %d, power off.", device->kernels[i]->hardware, device->statesStored[i]);
                 status = gckHARDWARE_SetPowerState(device->kernels[i]->hardware, gcvPOWER_OFF);
                 if (gcmIS_ERROR(status)) {
                     return -1;
@@ -1609,6 +1610,7 @@ static int viv_dev_resume(struct platform_device *dev)
                     break;
                 }
 
+                hae_print("hard: %p, store_state: %d, resume.", device->kernels[i]->hardware, statesStored);
                 /* Restore states. */
                 status = gckHARDWARE_SetPowerState(device->kernels[i]->hardware, statesStored);
                 if (gcmIS_ERROR(status)) {
diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/platform/eswin/gc_hal_kernel_platform_win2030.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/platform/eswin/gc_hal_kernel_platform_win2030.c
index d0bdc1a449e7..103fa3eceae0 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/platform/eswin/gc_hal_kernel_platform_win2030.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/platform/eswin/gc_hal_kernel_platform_win2030.c
@@ -115,7 +115,9 @@
 #endif
 #include <linux/devfreq.h>
 #include <linux/pm_opp.h>
+
 #define G2D_HILOAD_CLK 1040000000
+#define G2D_INIT_DEV_FREQ 0xffffffff
 
 /* Disable MSI for internal FPGA build except PPC */
 #if gcdFPGA_BUILD
@@ -246,54 +248,141 @@ static void show_clk_status(int dieIndex)
 }
 
 #if defined(CONFIG_PM_DEVFREQ)
-static int g2d_devfreq_target(struct device *dev, unsigned long *freq, u32 flags) {
+static unsigned long g_dev_cur_freq[gcdDEVICE_COUNT] = {G2D_INIT_DEV_FREQ, G2D_INIT_DEV_FREQ};
+
+static int g2d_devfreq_get_init_freq(struct device *dev)
+{
     int i, j;
-    unsigned long rate = *freq;
-    int ret = -1;
-    struct clk *aclk;
 
     for (i = 0; i < gpd.num_domains; i++) {
-        if (gpd.dev[i] != dev) continue;
+        if (gpd.dev[i] != dev) {
+            continue;
+        }
+
         for (j = 0; j < nc_of_clks; j++) {
-            if (strcmp("g2d_clk", clk_names[j]) && strcmp("g2d_aclk", clk_names[j])) continue;
-            aclk = gpd.clks[i][j];
-            rate = clk_round_rate(aclk, rate);
-            if (rate > 0) {
-                ret = clk_set_rate(aclk, rate);
-                if (ret) {
-                    dev_err(dev, "failed to set %s clk: %d\n", clk_names[j], ret);
-                    return ret;
-                }
+            if (!strcmp("g2d_clk", clk_names[j])) {
+                g_dev_cur_freq[i] = clk_get_rate(gpd.clks[i][j]);
+                return 0;
             }
-            dev_info(dev, "set %s rate to %ldHZ\n", clk_names[j], rate);
         }
     }
-    if (!ret) {
-        *freq = rate;
+
+    dev_err(dev, "not find g2d clk!\n");
+    return -1;
+}
+
+static int g2d_devfreq_get_cur_freq(struct device *dev, unsigned long *freq) {
+    int ret = -1;
+    int i;
+
+    for (i = 0; i < gpd.num_domains; i++) {
+        if (gpd.dev[i] == dev) {
+            continue;
+        }
+        if (g_dev_cur_freq[i] == G2D_INIT_DEV_FREQ) {
+            ret = g2d_devfreq_get_init_freq(dev);
+            if (ret != 0)  {
+                return -1;
+            }
+        }
+        *freq = g_dev_cur_freq[i];
         return 0;
-    } else {
-        dev_err(dev, "set rate to %ldHZ\n failed", rate);
-        return -1;
     }
+
+    return -1;
 }
-static int g2d_devfreq_get_cur_freq(struct device *dev, unsigned long *freq) {
+
+static int g2d_devfreq_target(struct device *dev, unsigned long *freq, u32 flags) {
     int i, j;
+    unsigned long round_rate = 0;
+    int ret = -1;
+    struct clk *clk_handle;
+
     for (i = 0; i < gpd.num_domains; i++) {
-        if (gpd.dev[i] != dev) continue;
+        if (gpd.dev[i] != dev) {
+            continue;
+        }
+
         for (j = 0; j < nc_of_clks; j++) {
-            if (!strcmp("g2d_clk", clk_names[j])) {
-                return clk_get_rate(gpd.clks[i][j]);
+            if (strcmp("g2d_clk", clk_names[j])) {
+                continue;
             }
+
+            clk_handle = gpd.clks[i][j];
+
+            round_rate = clk_round_rate(clk_handle, *freq);
+            if (round_rate <= 0) {
+                dev_err(dev, "failed to set %s clk: %d\n", clk_names[j], ret);
+                return -1;
+            }
+        }
+
+        if (round_rate != *freq) {
+            *freq = round_rate;
+        }
+
+        if (g_dev_cur_freq[i] == round_rate) {
+            return 0;
         }
+
+        ret = clk_set_rate(clk_handle, round_rate);
+        if (ret) {
+            dev_err(dev, "failed to set %s clk: %d\n", clk_names[j], ret);
+            return -1;
+        }
+
+        dev_info(dev, "sys set rate from %luHz to %luHz\n", g_dev_cur_freq[i], round_rate);
+
+        g_dev_cur_freq[i] = round_rate;
     }
+
     return 0;
 }
+
+static int g2d_get_dev_freq_status(struct device *dev, struct devfreq_dev_status *stat)
+{
+    int i;
+
+    if (dev == NULL || stat == NULL) {
+        pr_err("%s param invalid!\n", __func__);
+        return -1;
+    }
+
+    for (i = 0; i < gpd.num_domains; i++) {
+        if (gpd.dev[i] != dev) {
+            continue;
+        }
+
+        stat->current_frequency = g_dev_cur_freq[i];
+        stat->busy_time = 1024; // not support hardware load now.
+        stat->total_time = 1024; // 100
+        return 0;
+    }
+
+    dev_err(dev, "get hae dev freq failed!\n");
+    return -1;
+}
+
+static void eswin_exit(struct device *dev)
+{
+    return;
+}
+
+static struct devfreq_simple_ondemand_data ondemand_data =
+{
+    .upthreshold =80,
+    .downdifferential=10,
+};
+
 static struct devfreq_dev_profile g2d_devfreq_profile = {
     .initial_freq = G2D_HILOAD_CLK,
     .timer = DEVFREQ_TIMER_DELAYED,
     .polling_ms = 1000, /* Poll every 1000ms to monitor load */
     .target = g2d_devfreq_target,
     .get_cur_freq = g2d_devfreq_get_cur_freq,
+    .get_dev_status = g2d_get_dev_freq_status,
+    .exit = eswin_exit,
+    .is_cooling_device = true,
 };
 #endif
 
@@ -393,12 +482,17 @@ static int gpu_add_power_domains(struct platform_device *pdev, gcsMODULE_PARAMET
             hae_print("Failed to add OPP table");
             goto error;
         }
-        df = devm_devfreq_add_device(gpd.dev[i], &g2d_devfreq_profile, "userspace", NULL);
+        df = devm_devfreq_add_device(gpd.dev[i], &g2d_devfreq_profile, DEVFREQ_GOV_SIMPLE_ONDEMAND, &ondemand_data);
         if (IS_ERR(df)) {
             ret = PTR_ERR(df);
             hae_print("add devfreq failed");
             goto error;
         }
+        ret = devm_devfreq_register_opp_notifier(gpd.dev[i], df);
+        if (ret < 0) {
+            hae_print("i: %d, dev: %p, failed to register opp notifier\n", i, gpd.dev[i]);
+            goto error;
+        }
         gpd.df[i] = df;
 #endif
 
@@ -861,7 +955,7 @@ static struct _gcsPLATFORM default_platform = {
 gceSTATUS
 _AdjustParam(gcsPLATFORM *Platform, gcsMODULE_PARAMETERS *Args)
 {
-    int ret;
+    int ret = -1;
 #if gcdSUPPORT_DEVICE_TREE_SOURCE
     ret = gpu_parse_dt(Platform->device, Args);
     if(gcmIS_SUCCESS(ret)){
-- 
2.53.0

